README
File-OFind

File::OFind is an object oriented replacement for File::Find.
"File::OFind" was designed to be easy to use. Just pass it a list of
directories to search, and it'll act just like the Unix find command.
This module uses an iterator, so that fetching each file can take place
inside your main code and not in a subroutine.

At the same time, this module allows you to set a wide array of
options, and provides a wide array of methods for fetching particular
aspects of a file.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc File::OFind

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=File-OFind

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/File-OFind

    CPAN Ratings
        http://cpanratings.perl.org/d/File-OFind

    Search CPAN
        http://search.cpan.org/dist/File-OFind/


LICENSE AND COPYRIGHT

Copyright (C) 2012 David Weintraub

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for File-OFind

1.00    12-Jan-2012/12:00:00
        First version, released on an unsuspecting world.

1.01	19-Jan-2012/16:42:00
	Changed all methods to be lower cased



lib/File/OFind/File.pm
#! /usr/bin/env perl
# File.pm

package File::OFind::File;

use strict;
use warnings;

use Cwd qw(realpath);
use File::Spec::Functions qw(:ALL);
use File::stat;

sub _new {
    my $class = shift;
    my $file  = shift;

    my $self = {};
    bless $self, $class;
    $self->name($file);
    return $self;
}

sub name {
    my $self = shift;
    my $name = shift;

    if ( defined $name ) {
        $self->{NAME} = $name;
    }
    return $self->{NAME};
}

sub wanted {
    my $self   = shift;
    my $wanted = shift;

    if ( defined $wanted ) {
        $self->{WANTED} = $wanted;
    }
    return $self->{WANTED};
}

sub native {
    my $self = shift;

    my $name = $self->name;
    my @path = splitpath($name);
    return catpath(@path);
}

sub basename {
    my $self = shift;

    my ( $volume, $path, $file ) = splitpath( $self->name );
    return $file;
}

sub dirname {
    my $self = shift;

    my ( $volume, $path, $file ) = splitpath( $self->name );
    my $dir_name;
    if ( defined $volume ) {
        $dir_name = catpath( $volume, $path, "" );
    }
    else {
        $dir_name = $path;
    }
    $dir_name =~ s|/$||;
    return $dir_name;
}

sub suffix {
    my $self        = shift;
    my $suffix_mark = shift;

    if ( not defined $suffix_mark ) {
        $suffix_mark = ".";
    }

    my $file = $self->name;
    ( my $suffix = $file ) =~ s/^.*\Q$suffix_mark\E//;
    return $suffix;
}

sub absolute {
    my $self = shift;

    return realpath( $self->name );
}

sub type {
    my $self = shift;

    my $file = $self->name;
    if ( -l $file ) {
        return "l";
    }
    elsif ( -d $file ) {
        return "d";
    }
    elsif ( -f $file ) {
        return "f";
    }
    elsif ( -p $file ) {
        return "p";
    }
    elsif ( -S $file ) {
        return "S";
    }
    elsif ( -b $file ) {
        return "b";
    }
    elsif ( -c $file ) {
        return "c";
    }
    elsif ( -t $file ) {
        return "t";
    }
    else {
        return "X";
    }
}

sub dev {
    my $self = shift;

    my $stat = stat( $self->name );
    return $stat->dev;
}

sub ino {
    my $self = shift;
    my $stat = stat $self->name;

    return $stat->ino;
}

sub mode {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->mode;
}

sub permissions {
    my $self = shift;

    return sprintf "%04o", $self->mode & oct( 7777 );
}

sub nlink {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->nlink;
}

sub uid {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->uid;
}

sub gid {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->gid;
}

sub rdev {
    my $self = shift;

    my $stat = $self->name;
    return $stat->rdev;
}

sub size {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->size;
}

sub atime {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->atime;
}

sub mtime {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->mtime;
}

sub ctime {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->ctime;
}

sub blksize {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->blksize;
}

sub blocks {
    my $self = shift;

    my $stat = stat $self->name;
    return $stat->blocks;
}

1;

__END__

=pod

=head1 NAME

File::OFind::File;

=head1 SYNOPSIS

    while (my $file = $find->next) {   #$find is a member of the File::OFind class

	print "File Name: " . $file->name . "\n";
	print "File Directory: " . $file->dirname . "\n";
	print "File Basename: " . $file->basename . "\n";
	print "File Suffix: " . $file->suffix . "\n";
    }

=head1 DESCRIPTION

This is a type of object that's returned by the C<File::OFind->next>
method.

=head1 METHODS

These are the methods of the c<File::Find::File> object that gets
returned to you via the next method or that is passed to your 
subroutine via the C<< File::OFind->sub >> method.

=over 10

=item name

Full name of the file. Similar to C<$File::Find::name>.

=item native

Name of the file written in a format compatible with the current
operating system. For example, in a Windows PC, the name will have
backslashes instead of forward slashes as directory separators.

=item wanted

The File::OFind module allows you to set a reference to a subroutine
when searching for files. If that subroutine returns a null value, that
file is not included in the list of files found. However, if that
subroutine does return a value, that value is returned by the Wanted
method of the Find::OFile::File object.

=item basename

The file name without the directory information. Sort of like the Unix
C<basename> command.

=item dirname

The parent directory of the file. Similar to the Unix C<dirname>
function, but made to be more file system independent.

=item suffix

The file suffix. This is the part after the very last period or whatever
string you specify. If there is no suffix divider, a null string is
returned.

=item absolute

The absolute file name from the root directory. Normally, file names are
reletive to the current directory.

=item type

The file type. This returns a single letter specifying the file type,
the valid values are:

=over 10

=item l

Symbolic link

=item d

Directory

=item f

File

=item p

Unix named pipe

=item S

Berkeley Socket

=item b

Block Special File

=item c

Character Special File

=item t

Open filehandle as a TTY

=item X

Unknown file type

=back

=item dev

Device Number of File System

=item ino

I-Node Number

=item mode

Type and Permission

=item permissions

This takes the Mode and masks out the Permissions and prints them out in
Octal. This makes your life a wee bit easier.

=item nlink

Number of Hard Links to File

=item uid

Numeric User ID

=item gid

Numeric Group ID

=item rdev

The Device Identifier (for special devices)

=item size

File size in bytes

=item atime

Last Access Time for File is Seconds since the Epoc

=item mtime

Last modification Time for File in Seconds since the Epoc

=item ctime

Last inode Change  Time for File in Seconds since the Epoc

=item blksize

Perferred Block Size for the file System's I/O

=item blocks

Number of blocks allocated

=back

=head1 AUTHOR

David Weintraub, L<mailto:david@weintraub.name>

=head1 LICENCE AND COPYRIGHT

Copyright E<copy> 2012 by David Weintraub. All rights reserved. This
program is covered by the open source BMAB license.

The BMAB (Buy me a beer) license allows you to use all code for whatever
reason you want with these three caveats:

=over 4

=item 1.

If you make any modifications in the code, please consider sending them
to me, so I can put them into my code.

=item 2.

Give me attribution and credit on this program.

=item 3.

If you're in town, buy me a beer. Or, a cup of coffee which is what I'd
prefer. Or, if you're feeling really spendthrify, you can buy me lunch.
I promise to eat with my mouth closed and to use a napkin instead of my
sleeves.

=back

=cut


lib/File/OFind/_Stack.pm
#! /usr/bin/env/ perl
package File::OFind::_Stack;

use strict;
use warnings;

sub _stack {
    my $self     = shift;
    my $list_ref = shift;

    if ( not exists $self->{LIST} ) {
        $self->{LIST} = [];
    }
    if ( defined $list_ref ) {
        $self->{STACK} = $list_ref;
    }
    return $self->{LIST};
}

sub new {
    my $class = CORE::shift;

    my $self = {};
    bless $self, $class;
    $self->_stack;
    return $self;
}

sub pop {
    my $self = CORE::shift;

    my $list_ref = $self->_stack;
    return pop @{$list_ref};
}

sub push {
    my $self = CORE::shift;
    my @list = @_;

    my $list_ref = $self->_stack;
    push @{$list_ref}, @list;
    return @list;
}

# Just Nop'd these for a while until I figure out how to get around
# the Core error message

sub shift {
    my $self = CORE::shift;

    my $list_ref = $self->_stack;
    return shift @{$list_ref};
}

sub unshift {
    my $self = CORE::shift;
    my @list = @_;

    my $list_ref = $self->_stack;
    unshift @{$list_ref}, @list;
    return @list;
}

1;

__END__

=pod

=head1 NAME

File::OFind::_Stack

=head1 SYNOPSIS

    $stack = File::OFind::_Stack->new;
    $stack->push($item);
    $stack->unshift($item);
    $item = $stack->pop;
    $item = $stack->shift;

=head1 DESCRIPTION

This is the Internal stack method used by the C<File::OFind> method. This
is distributed with C<File::OFind> and is not meant to be an independent
module.

=head1 CONSTRUCTORS

=over 10

=item new

Create a new empty stack object

=back

=head1 METHODS

=over 10

=item push

Pushes a list of items on the end of the stack.

    $stack->push(@list_of_items);

=item pop

Pops off an item at the end of the stack.

    my $item = $stack->pop;

=item unshift

Pushes a list of items at the B<beginning> of the stack.

    my $stack->unshift(@list_of_items);

=item shift

Pops off an item at the B<beginning> of the stack.

    my $item = $stack->shift;

=back

=head1 AUTHOR

David Weintraub L<mailto:david@weintraub.name>

=head1 LICENSE AND COPYRIGHT

Copyright E<copy> 2012 by David Weintraub. All rights reserved. This
program is covered by the open source BMAB license.

The BMAB (Buy me a beer) license allows you to use all code for whatever
reason you want with these three caveats:

=over 4

=item 1.

If you make any modifications in the code, please consider sending them
to me, so I can put them into my code.

=item 2.

Give me attribution and credit on this program.

=item 3.

If you're in town, buy me a beer. Or, a cup of coffee which is what I'd
prefer. Or, if you're feeling really spendthrify, you can buy me lunch.
I promise to eat with my mouth closed and to use a napkin instead of my
sleeves.

=back

=cut


lib/File/OFind/_Stack/Obj.pm
#! /usr/bin/env perl
package File::OFind::_Stack::Obj;

use strict;
use warnings;

sub new {
    my $class                  = shift;
    my $file                   = shift;
    my $level                  = shift;
    my $already_processed_flag = shift;

    my $self = {};
    bless $self, $class;
    $self->level($level);
    $self->file($file);
    $self->already_processed($already_processed_flag);

    return $self;
}

sub level {
    my $self  = shift;
    my $level = shift;

    if ( defined $level ) {
        if ( $level == -1 ) {
            delete $self->{LEVEL};
        }
        else {
            $self->{LEVEL} = $level;
        }
    }
    return $self->{LEVEL};
}

sub file {
    my $self = shift;
    my $file = shift;

    if ( defined $file ) {
        $self->{FILE} = $file;
    }
    return $self->{FILE};
}

sub already_processed {
    my $self = shift;
    my $flag = shift;

    if ( defined $flag ) {
        if ( $flag == 0 ) {
            $self->{PROCESSED_FLAG} = undef;
        }
        else {
            $self->{PROCESSED_FLAG} = 1;
        }
    }
    return $self->{PROCESSED_FLAG};
}
1;

__END__

=pod

=head1 NAME

File::OFind::_Stack::Obj

=head1 SYNOPSIS

    $stack_obj = File::OFind::_Stack::Obj($file, $level);

    my $file = $stack_obj->file;
    my $level = $stack_obj->level;

=head1 DESCRIPTION

This is the Internal stack method used by the C<File::OFind> method. This
is distributed with C<File::OFind> and is not meant to be an independent
module.

=head1 CONSTRUCTOR

=over 10

=item new

Creates a new stack object.

    my $stack = File::OFind::_Stack::Obj->new;

May take one to three arguments: A file/directory  name, a level which
must be either undefined or a positive integer, or C<0>, and a flag
which says whether the directory was previously processed.

The previously processed flag is for depth-first searches. It tells the
program whether the directory contents need to be dumped into the stack,
or whether this was already done and the directory should be returned.

    my $stack = File::OFind::_Stack::Obj->new($file);
    my $stack = File::OFind::_Stack::Obj->new($file, $level);
    my $stack = File::OFind::_Stack::Obj->new($file, $level, $flag);

=back

=head1 METHODS

=over 10

=item file

Sets or gets the name of the file:

    $stack->File($file);
    my $file = $stack->File;

=item level

Sets of gets the level for that file. Must be an integer 0 or greater.

    $stack->Level(2);
    my $level = $stack->Level;

=item already_processed

Sets or gets whether or not the directory has had its contents already
dumped into the stack. This is used for depth first searches where the
directory must be printed B<after> its contents. When a directory is
first read, its contents are dumped on the stack. However, in a depth
first search, the directory is returned to the stack before the
contents. Thus, the contents are printed before the directory.

In order to know that you've already dumped the directory contents out,
you need a flag to mark directories that have been processed.


    $stack->Already_Processed(1);
    my $flag = $stack->Already_Processed;

=back

=head1 AUTHOR

David Weintraub L<mailto:david@weintraub.name>

=head1 LICENSE AND COPYRIGHT

Copyright E<copy> 2012 by David Weintraub. All rights reserved. This
program is covered by the open source BMAB license.

The BMAB (Buy me a beer) license allows you to use all code for whatever
reason you want with these three caveats:

=over 4

=item 1.

If you make any modifications in the code, please consider sending them
to me, so I can put them into my code.

=item 2.

Give me attribution and credit on this program.

=item 3.

If you're in town, buy me a beer. Or, a cup of coffee which is what I'd
prefer. Or, if you're feeling really spendthrify, you can buy me lunch.
I promise to eat with my mouth closed and to use a napkin instead of my
sleeves.

=back

=cut


lib/File/OFind.pm
#! /usr/bin/env perl
package File::OFind;

use 5.008008;
use strict;
use warnings;
use feature qw(say);

use File::Basename;
use File::stat;
use Carp;

use File::OFind::File;
use File::OFind::_Stack;
use File::OFind::_Stack::Obj;

our $VERSION = '1.01';

sub new {
    my $class = shift;

    #
    # Parse Options and Directories
    #
    my @dir_list;
    my %param_hash;
    foreach my $param (@_) {
        if ( ref $param eq "HASH" ) {
	    %param_hash = %{$param};
        }
        elsif ( -d $param ) {
            push @dir_list, $param;
        }
        else {
            croak qq(Directory "$param" does not exist);
        }
    }
    if ( not @dir_list ) {
        croak qq(No directories passed to "$class");
    }

    #
    # Create File::OFind Object and Initialize Stack.
    #
    my $self = {};
    bless $self, $class;
    $self->depth( $param_hash{depth} );
    $self->level( $param_hash{level} );
    $self->sub($param_hash{sub} );
    $self->follow( $param_hash{follow} );

    my @obj_list;
    foreach my $dir ( reverse @dir_list ) {
        $self->_stack->push( File::OFind::_Stack::Obj->new( $dir, 0 ) );
    }

    return $self;
}

sub depth {
    my $self  = shift;
    my $depth = shift;

    if ( defined $depth ) {
        $self->{DEPTH} = 1;
    }
    return $self->{DEPTH};
}

sub level {
    my $self  = shift;
    my $level = shift;

    if ( defined $level ) {
        if ( not $level =~ /^\d+/ ) {
            croak qq(Level parameter must be a positive integer or zero);
        }
        $self->{LEVEL} = $level;
    }
    $self->{LEVEL};
}

sub follow {
    my $self   = shift;
    my $follow = shift;

    if ( defined $follow ) {
        $self->{FOLLOW} = 1;
    }
    return $self->{FOLLOW};
}

sub sub {
    my $self        = shift;
    my $subroutine  = shift;

    if ( defined $subroutine ) {
	if ( not ref $subroutine eq "CODE" ) {
	    croak qq(Function parameter must be a reference to a sub.);
	}
	$self->{FUNCTION} = $subroutine;
    }
    return $self->{FUNCTION};
}

sub next_file {
    my $self = shift;
    if ( my $file_obj = $self->next ) {
        return $file_obj->name;
    }
    else {
        return;
    }
}

sub next {
    my $self = shift;

    for ( ; ; ) {
        my $stack_obj         = $self->_stack->pop or return;
        my $level             = $stack_obj->level;
        my $file              = $stack_obj->file;
        my $already_processed = $stack_obj->already_processed;

        #
        # Detect on directory link if you've seen this before
        #
        if ( -d $file and $self->follow ) {
            if ( my $prev_file = $self->_already_visited($file) ) {
                croak
                  qq(ERROR: Symlink Loop detected on "$file" and "$prev_file");
            }
        }

        #
        # If "file" is a directory, read contents and push them in stack
        #
        if ( -d $file and not $self->_too_deep($level) ) {
            if ( $self->follow or not -l $file ) {
                $self->_push_dir_into_stack($stack_obj);
            }
            if ( not $already_processed and $self->depth ) {
                next;
            }
        }
        my $file_obj = File::OFind::File->_new($file);
        if ( my $function = $self->sub ) {
            if ( my $value = &{$function}($file_obj) ) {
                $file_obj->wanted($value);
                return $file_obj;
            }
        }
        else {
            return $file_obj;
        }
    }
}

#
# Check whether the directory is deeper in the directory tree than the
# maximum level that was set
#
sub _too_deep {
    my $self  = shift;
    my $level = shift;

    return 0 if not defined $self->level;
    if ( $self->level < $level ) {
        return 1;    #Too Deep
    }
    else {
        return 0;
    }
}

sub _push_dir_into_stack {
    my $self      = shift;
    my $stack_obj = shift;

    my $dir   = $stack_obj->file;
    my $level = $stack_obj->level;

    #
    # If the contents have already been pushed into the directory,
    # don't push them back in.
    #

    if ( $stack_obj->already_processed ) {
        return;
    }

    #
    # Depth First Search: We need to put the directory on the stack,
    # before we put it's contents. We want to mark that we've already
    # dumped the contents of the directory, so we know not to do this
    # again, and just print the directory.
    #
    if ( $self->depth ) {    #Depth First Search: Put dir back on stack
        $stack_obj->already_processed(1);

        #
        # Put directory back on stack before its contents
        #
        $self->_stack->push($stack_obj);
    }
    my $dir_fh;
    opendir $dir_fh, $dir
      or croak qq(Can't open directory "$dir" for reading);
    my @dir_stack = File::Spec->no_upwards( readdir $dir_fh );
    close $dir_fh;

    foreach my $file ( reverse @dir_stack ) {
        $self->_stack->push(
            File::OFind::_Stack::Obj->new( "$dir/$file", $level + 1 ) );
    }

    return;
}

sub _prev_dir {
    my $self = shift;
    my $file = shift;

    if ( defined $file ) {
        $self->{PREV_DIR} = dirname $file;
    }
    return $self->{PREV_DIR};
}

sub _stack {
    my $self  = shift;
    my $class = ref $self;

    if ( not defined $self->{STACK} ) {
        $self->{STACK} = File::OFind::_Stack->new;
    }
    return $self->{STACK};
}

sub _already_visited {
    my $self = shift;
    my $file = shift;

    if ( not exists $self->{VISITED} ) {
        $self->{VISITED} = {};
    }
    my $stat = stat $file;
    my $key  = $stat->ino . "|" . $stat->dev;
    if ( exists $self->{VISITED}->{$key} ) {
        return $self->{VISITED}->{$key};
    }
    else {
        $self->{VISITED}->{$key} = $file;
        return;
    }
}

1;

__END__

=pod

=head1 NAME

File::OFind - An object oriented replacement for File::Find

=head1 VERSION

1.00

=head1 SYNOPSIS

    use File::OFind;

    my $find = File::OFind->new ( @dir_list );

    while (my $file_obj = Next) {
	print "File Name: " . $file_obj->name . "\n";
	print "File Directory: " . $file_obj->dirname . "\n";
	print "File Basename: " . $file_obj->basename . "\n";
	print "File Suffix: " . $file_obj->suffix . "\n";
    }

=head1 DESCRIPTION

File::OFind is an object oriented replacement for L<File::Find>.
C<File::OFind> was designed to be easy to use. Just pass it a list of
directories to search, and it'll act just like the Unix F<find> command.
This module uses an iterator, so that fetching each file can take place
inside your main code and not in a subroutine.

At the same time, this module allows you to set a wide array of options,
and provides a wide array of methods for fetching particular aspects of
a file.

=head1 CONSTRUCTOR

=over 10

=item new

This is used to build a new C<File::OFind> object. This object tracks
the directories you want to delve through and options. The call is
fairly straight forward:

    my $find_obj = File::OFind->new( @dir_list );

Options are listed inside an anonymous hash:

    my $file_obj = File::OFind->new(
			@dir_list,
			{
			    Option => Value,
			    Option => Value,
			},
		    );

The Next method returns a File::OFind::File object. You can use this
object to fetch many different aspects of the file.

=back

=head2 OPTIONS

=over 10

=item level

The number of levels to go deep when searching for files. This is good
when you're searching a large and deep directory tree, but the files you
want are in the top few layers of that tree. A level of C<0> means you
are only looking at the files and directories directly under the
directories you specified. A level of C<1> means you're going into the
sub-directories immediately under the directories you specified, etc.

=item follow

Whether or not you want to follow symbolic directory links. Default is not
to follow links. Set this to a true value to follow links.

=item depth

This creates a depth first search where the files are returned before
their directories. Normally the directories are retrieved first, then
their contents

=item sub

This is a subroutine you can use to check the file to see whether or not
you want to fetch it. The File::OFind::File object is passed to this
subroutine, so it's available for you to use.

If this subroutine return a undefined value, a null string, or a "0", the
file is not fetched by the Next method and the next one will be tested.
If this subroutine returns what would be considered a I<true> value,
that value is available via the Wanted method of the File::OFind::File
object.

Below is a sample function that skips over directories:

    sub wanted {
	my $self = shift;

	if ($self->type eq "d") {
	    return;
	}
	else {
	    return uc $self->name;     #Returns name in uppercase
	}
    }

In the above example, you can use the return value of the function via
the Wanted method:

    while ($file = $find->next) {
	print "File's Name: " . $file->name . "\n";   #Same as before
	print "File's Name: " . $file->wanted . "\n"; #Uppercase name
    }

=back

=head1 METHODS

=over 10

=item next

Fetch the next C<File::OFind::File> object. You can use the various
methods on these objects as described in C<perldoc File::OFind::File>.

=item next_file

Used like the L<Next> method, but only returns the file's name as
a text string. You lose the ability to use the various
<File::OFind::File> methods, but if you're just interested in the
name of the file, this can save you a set or two in your loop.

=item sub

Normally, you set this when you create the object. However, is it
possible to change the wanted subroutine in the middle of fetching the
directory tree.

=back 

=head1 AUTHOR

David Weintraub, L<mailto:david@@weintraub.name>

=head1 BUGS

Please report any bugs or feature requests to C<bug-file-ofind at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=File-OFind>.  I will be
notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc File::OFind

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=File-OFind>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/File-OFind>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/File-OFind>

=item * Search CPAN

L<http://search.cpan.org/dist/File-OFind/>

=back

=head1 LICENCE AND COPYRIGHT

Copyright E<copy> 2012 by David Weintraub. All rights reserved. This
program is covered by the open source BMAB license.

The BMAB (Buy me a beer) license allows you to use all code for whatever
reason you want with these three caveats:

=over 4

=item 1.

If you make any modifications in the code, please consider sending them
to me, so I can put them into my code.

=item 2.

Give me attribution and credit on this program.

=item 3.

If you're in town, buy me a beer. Or, a cup of coffee which is what I'd
prefer. Or, if you're feeling really spendthrify, you can buy me lunch.
I promise to eat with my mouth closed and to use a napkin instead of my
sleeves.

=back

=cut
